waiter/src/waiter/auth/oauth.clj

```diff
@@ -0,0 +1,166 @@
+                         query-string')]
+    (str \/ path query-string'' hash)))
+
+(defn provider-list-handler
+  "Responds with a login page listing available OAuth providers."
+  [providers {:keys [query-string request-method] {:keys [path]} :route-params}]
+  (case request-method
+    :get {:body (template/eval (slurp (io/resource "web/oauth.html"))
+                               {:providers providers
+                                :display-name (fn [provider] (display-name provider))
+                                :hash-query-param-name HASH-QUERY-PARAM-NAME
+                                :path path
+                                :query-string query-string})
+          :headers {"content-type" "text/html"}}
+    (throw (ex-info "Invalid request method"
+                    {:request-method request-method
+                     :status 405}))))
+
+(defn match-oauth-route
+  "Match an OAuth route based upon a uri."
+  [uri]
+  (let [route-map ["/waiter-auth/oauth" {["/providers/" [#".*" :path]] :provider-list
+                                         "/" {[:provider "/redirect/" [#".*" :path]] :redirect
+                                              [:provider "/authenticate"] :authenticate}}]]
+    (bidi/match-route route-map uri)))
+
+(defn wrap-oauth
+  "Middleware for handling OAuth."
+  [handler-fn run-as-user password providers]
+  (fn [{:keys [headers query-string uri] :as request}]
+    (try
+      (let [{:keys [handler route-params] :as match} (match-oauth-route uri)
+            request' (assoc request :route-params route-params)
+            provider-fn (fn [] (-> route-params :provider keyword providers))]
+        (case handler
+          :provider-list (provider-list-handler providers request')
+          :redirect (redirect (provider-fn) request')
+          :authenticate (authenticate (provider-fn) request')
+          ; else
+          (let [waiter-cookie (auth/get-auth-cookie-value (get headers "cookie"))
+                [auth-principal _ :as decoded-auth-cookie] (auth/decode-auth-cookie waiter-cookie password)]
+            (if (auth/decoded-auth-valid? decoded-auth-cookie)
+              (-> request
+                  (auth/assoc-auth-in-request auth-principal)
+                  handler-fn)
+              {:status 307
+               :headers {"location" (cond-> (str "/waiter-auth/oauth/providers" uri)
+                                      (not (str/blank? query-string)) (str \? query-string))}}))))
+      (catch Exception ex
+        (utils/exception->response ex request)))))
+
+(defrecord OAuthAuthenticator [password providers run-as-user]
+  auth/Authenticator
+  (auth-type [_]
+    :oauth)
+  (check-user [_ user _]
+    user)
+  (wrap-auth-handler [_ handler]
+    (wrap-oauth handler run-as-user password providers)))
+
+(defn oauth-authenticator
+  "Factory function for creating OAuthAuthenticator"
+  [{:keys [password providers run-as-user]}]
+  {:pre [password providers run-as-user]}
```

Comment:
The example provided in `config-shell-oauth.edn` does not have a password entry.

Reply:
Should it be an error if providers is empty?

Reply:
The password is passed in via context:

<pre>
 :authenticator (pc/fnk [[:settings authenticator-config]
                           passwords]
                    (utils/create-component authenticator-config :context {:password (first passwords)}))
</pre>